### PR TITLE
[Feature][Core] Upgrade flink source translation

### PR DESCRIPTION
**Pull request by TyrantLucifer**
### Purpose of this pull request

Refactor flink translation layer using flink new source api that come from FLIP-27

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?

Origin E2E test cases is enough.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by TyrantLucifer** [comment]
@hailin0 @Hisoka-X @EricJoy2048 PTAL

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java [hunk 1]
```diff
@@ -32,8 +35,11 @@
 @Slf4j
 public class FileSourceSplitEnumerator
         implements SourceSplitEnumerator<FileSourceSplit, FileSourceState> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileSourceSplitEnumerator.class);
+
```
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java:37]: Is it equivalent to @Slf4j？

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml [hunk 2]
```diff
@@ -72,39 +72,46 @@
...
+            <version>${flink.1.15.3.version}</version>
         </dependency>
+
         <dependency>
             <groupId>org.apache.flink</groupId>
-            <artifactId>flink-clients_${scala.binary.version}</artifactId>
-            <version>${flink.1.13.6.version}</version>
-            <scope>${flink.scope}</scope>
+            <artifactId>flink-table-planner-loader</artifactId>
+            <version>1.15.3</version>
```
Comment (Carl-Zhou-CN) [inline seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]:
```suggestion
            <version>${flink.1.15.3.version}</version>
```

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml [hunk 3]
```diff
@@ -72,39 +72,46 @@
...
         </dependency>
+
         <dependency>
             <groupId>org.apache.flink</groupId>
-            <artifactId>flink-runtime-web_${scala.binary.version}</artifactId>
-            <version>${flink.1.13.6.version}</version>
-            <scope>${flink.scope}</scope>
+            <artifactId>flink-table-runtime</artifactId>
+            <version>1.15.3</version>
```
Comment (Carl-Zhou-CN) [inline seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: Same as above

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml [hunk 4]
```diff
@@ -72,39 +72,46 @@
...
         </dependency>
+
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-clients</artifactId>
+            <version>1.15.3</version>
```
Comment (Carl-Zhou-CN) [inline seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: Same as above

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSource.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Override
+    public SourceReader<Row, SplitWrapper<SplitT>> createReader(SourceReaderContext readerContext)
+            throws Exception {
+        org.apache.seatunnel.api.source.SourceReader.Context context =
+                new FlinkSourceReaderContext(readerContext, source);
+        org.apache.seatunnel.api.source.SourceReader<SeaTunnelRow, SplitT> reader =
+                source.createReader(context);
+        return new FlinkSourceReader<>(reader, (SeaTunnelRowType) source.getProducedType());
```
Comment (Carl-Zhou-CN) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSource.java]: source.getProducedType() ，have been marked as @Deprecated

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSourceSplitEnumeratorContext.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public void sendEventToSourceReader(int subtaskId, SourceEvent event) {
+        enumContext.sendEventToSourceReader(subtaskId, new SourceEventWrapper(event));
+    }
+
+    @Override
+    public MetricsContext getMetricsContext() {
+        return null;
```
Comment (Carl-Zhou-CN) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSourceSplitEnumeratorContext.java]:
```suggestion
        return new AbstractMetricsContext() {};
```
Would this be better?

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSource.java (hunk 5 above)
Comment (TyrantLucifer) [reply seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSource.java]: Current any connectors will not implement the multi table read/write feature, In this modification, we will temporarily call the original method first, next stage I will update this.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java (hunk 1 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java:37]: done

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml (hunk 2 above)
Comment (TyrantLucifer) [reply seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: done

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml (hunk 3 above)
Comment (TyrantLucifer) [reply seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: done

File: seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml (hunk 4 above)
Comment (TyrantLucifer) [reply seatunnel-examples/seatunnel-flink-connector-v2-example/pom.xml]: done

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSourceSplitEnumeratorContext.java (hunk 6 above)
Comment (TyrantLucifer) [reply seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSourceSplitEnumeratorContext.java]: done

**Review by Carl-Zhou-CN — DISMISSED** [review]
LGTM

File: seatunnel-connectors-v2/connector-amazondynamodb/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazondynamodb/source/AmazonDynamoDBSourceReader.java [hunk 7]
```diff
@@ -83,16 +84,24 @@ public void close() {
...
+            AmazonDynamoDBSourceSplit split = pendingSplits.poll();
+            if (split == null) {
+                log.info(
+                        "AmazonDynamoDB Source Reader [{}] waiting for splits",
+                        context.getIndexOfSubtask());
+                if (noMoreSplit) {
+                    // signal to the source that we have reached the end of the data.
+                    log.info("Closed the bounded amazonDynamodb source");
+                    context.signalNoMoreElement();
+                    Thread.sleep(2000L);
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-amazondynamodb/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazondynamodb/source/AmazonDynamoDBSourceReader.java:98]: Why need `Thread.sleep(2000L);` ?

File: seatunnel-connectors-v2/connector-amazondynamodb/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazondynamodb/source/AmazonDynamoDBSourceSplitEnumerator.java [hunk 8]
```diff
@@ -103,6 +100,7 @@ private void assignSplit(Set<Integer> readers) {
                     pendingSplits.put(reader, assignmentForReader);
                 }
             }
+            enumeratorContext.signalNoMoreSplits(reader);
         }
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-amazondynamodb/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazondynamodb/source/AmazonDynamoDBSourceSplitEnumerator.java:104]: This method will be invoked when restore too. https://github.com/apache/seatunnel/pull/5100/files#diff-50ae2032ed74589f0bf5dbfca6e5ad7ace518c5647e8b844bf343c7ea0ae3ad1L143 . So I think move `enumeratorContext.signalNoMoreSplits(reader);` to here not a good way.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java [hunk 9]
```diff
@@ -54,15 +57,18 @@ public FileSourceSplitEnumerator(
...
+        this.pendingSplit.addAll(discoverySplits());
     }
 
     @Override
     public void run() {
-        // do nothing
+        for (int i = 0; i < context.currentParallelism(); i++) {
+            LOGGER.info("Assigned splits to reader [{}]", i);
+            assignSplit(i);
+        }
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java:68]: when invoke `run`, does not mean the reader already started. So why we need do it after `registerReader`.

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/AbstractSingleSplitReader.java [hunk 10]
```diff
@@ -25,6 +26,23 @@
 
 public abstract class AbstractSingleSplitReader<T> implements SourceReader<T, SingleSplit> {
 
+    protected final Object lock = new Object();
+
+    protected volatile boolean noMoreSplits = false;
+
+    @Override
+    public void pollNext(Collector<T> output) throws Exception {
+        synchronized (lock) {
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/AbstractSingleSplitReader.java:35]: Why we need a lock in here?

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java [hunk 11]
```diff
@@ -76,24 +67,14 @@ public List<DataStreamTableInfo> execute(List<DataStreamTableInfo> upstreamDataS
...
 
-            DataStreamSource<Row> sourceStream =
-                    addSource(
-                            executionEnvironment,
-                            sourceFunction,
-                            "SeaTunnel " + internalSource.getClass().getSimpleName(),
-                            bounded);
+            DataStreamSource sourceStream =
+                    executionEnvironment.fromSource(
+                            flinkSource, WatermarkStrategy.noWatermarks(), "st-source");
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java]: The name too simple, not clear than old one.

File: seatunnel-connectors-v2/connector-amazondynamodb/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazondynamodb/source/AmazonDynamoDBSourceReader.java (hunk 7 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-amazondynamodb/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazondynamodb/source/AmazonDynamoDBSourceReader.java:98]: avoid read data duplicated

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java (hunk 9 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java:68]: Flink translation layer guarantee this method must be invoked after all reader registed.

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/AbstractSingleSplitReader.java (hunk 10 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/AbstractSingleSplitReader.java:35]: Considering the possibility of concurrent calls, locking would be safer.

File: seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java (hunk 11 above)
Comment (TyrantLucifer) [reply seatunnel-core/seatunnel-flink-starter/seatunnel-flink-starter-common/src/main/java/org/apache/seatunnel/core/starter/flink/execution/SourceExecuteProcessor.java]: done

File: seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSourceReaderContext.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    public org.apache.seatunnel.api.source.Boundedness getBoundedness() {
+        return source.getBoundedness();
+    }
+
+    @Override
+    public void signalNoMoreElement() {
+        readerContext.sendSourceEventToCoordinator(
+                new NoMoreElementEvent(readerContext.getIndexOfSubtask()));
+        try {
+            Thread.sleep(2000L);
```
Comment (hailin0) [inline seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-common/src/main/java/org/apache/seatunnel/translation/flink/source/FlinkSourceReaderContext.java]: why sleep？

**Review by dijiekstra — APPROVED** [review]
LGTM

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java (hunk 9 above)
Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/split/FileSourceSplitEnumerator.java:68]: You are right. Seem CI passed, ignore this one.

**Comment by Carl-Zhou-CN** [comment]
Overall LGTM +1

**Review by Carl-Zhou-CN — APPROVED** [review]
+1

**Pull request merged**